File: src/packages/resolver.test.ts
import { test } from "node:test";

import { expect } from "expect";

import {
  InMemoryPackageSource,
  latestVersion,
  resolveTransitive,
  searchPackages,
} from "./resolver.ts";
import { type Coordinates, coordinatesToString, type PackageMetadata } from "./types.ts";

function pkg(
  spec: string,
  dependencies: (string | { spec: string; scope?: string; optional?: boolean })[] = [],
): PackageMetadata {
  const parse = (s: string): Coordinates => {
    const [groupId, artifactId, version] = s.split(":");
    return { groupId: groupId!, artifactId: artifactId!, version: version! };
  };
  return {
    coordinates: parse(spec),
    dependencies: dependencies.map(d =>
      typeof d === "string" ? parse(d) : { ...parse(d.spec), scope: d.scope, optional: d.optional },
    ),
  };
}

const c = (spec: string): Coordinates => pkg(spec).coordinates;

test("a transitive chain resolves depth-first packages in breadth-first order", async () => {
  const source = new InMemoryPackageSource("test", [
    pkg("org.a:a:1", ["org.b:b:1"]),
    pkg("org.b:b:1", ["org.c:c:1"]),
    pkg("org.c:c:1"),
  ]);
  const resolution = await resolveTransitive([c("org.a:a:1")], [source]);
  expect(resolution.packages.map(p => coordinatesToString(p.coordinates))).toEqual([
    "org.a:a:1",
    "org.b:b:1",
    "org.c:c:1",
  ]);
  expect(resolution.packages.map(p => p.depth)).toEqual([0, 1, 2]);
  expect(resolution.packages[1]!.requestedBy).toEqual(c("org.a:a:1").valueOf());
  expect(resolution.conflicts).toEqual([]);
  expect(resolution.missing).toEqual([]);
});

test("onResolve fires once per resolved package, in discovery order", async () => {
  const source = new InMemoryPackageSource("test", [
    pkg("org.a:a:1", ["org.b:b:1", "org.c:c:1"]),
    pkg("org.b:b:1", ["org.c:c:1"]), // c reached twice but resolved once
    pkg("org.c:c:1"),
    pkg("org.x:x:1"), // a missing dep is still reported (we attempt it)
  ]);
  const seen: string[] = [];
  const resolution = await resolveTransitive(
    [c("org.a:a:1"), c("org.missing:m:1")],
    [source],
    coordinates => seen.push(coordinatesToString(coordinates)),
  );
  // one notification per UNIQUE package (c only once), including the missing one
  expect(seen).toEqual(["org.a:a:1", "org.missing:m:1", "org.b:b:1", "org.c:c:1"]);
  // and it matches what actually got resolved + attempted
  expect(seen.length).toBe(resolution.packages.length + resolution.missing.length);
});

test("nearest version wins a diamond conflict; the loser is recorded", async () => {
  const source = new InMemoryPackageSource("test", [
    pkg("org.a:a:1", ["org.b:b:1", "org.c:c:1"]),
    pkg("org.b:b:1"),
    pkg("org.b:b:2"),
    pkg("org.c:c:1", ["org.b:b:2"]), // farther from the root than a's direct b:1
  ]);
  const resolution = await resolveTransitive([c("org.a:a:1")], [source]);
  const names = resolution.packages.map(p => coordinatesToString(p.coordinates));
  expect(names).toContain("org.b:b:1");
  expect(names).not.toContain("org.b:b:2");
  expect(resolution.conflicts).toEqual([
    { key: "org.b:b", selected: "1", rejected: "2", rejectedBy: c("org.c:c:1") },
  ]);
});

test("dependency cycles terminate", async () => {
  const source = new InMemoryPackageSource("test", [
    pkg("org.a:a:1", ["org.b:b:1"]),
    pkg("org.b:b:1", ["org.a:a:1"]),
  ]);
  const resolution = await resolveTransitive([c("org.a:a:1")], [source]);
  expect(resolution.packages).toHaveLength(2);
});

test("test-scoped and optional dependencies do not propagate", async () => {
  const source = new InMemoryPackageSource("test", [
    pkg("org.a:a:1", [
      { spec: "org.t:t:1", scope: "test" },
      { spec: "org.o:o:1", optional: true },
      { spec: "org.r:r:1", scope: "runtime" },
    ]),
    pkg("org.r:r:1"),
  ]);
  const resolution = await resolveTransitive([c("org.a:a:1")], [source]);
  expect(resolution.packages.map(p => coordinatesToString(p.coordinates))).toEqual([
    "org.a:a:1",
    "org.r:r:1",
  ]);
  expect(resolution.missing).toEqual([]);
});

test("sources are consulted in order; later sources fill gaps", async () => {
  const primary = new InMemoryPackageSource("primary", [pkg("org.a:a:1", ["org.b:b:1"])]);
  const fallback = new InMemoryPackageSource("fallback", [pkg("org.a:a:1"), pkg("org.b:b:1")]);
  const resolution = await resolveTransitive([c("org.a:a:1")], [primary, fallback]);
  expect(resolution.packages.map(p => [coordinatesToString(p.coordinates), p.source])).toEqual([
    ["org.a:a:1", "primary"], // primary wins for a (and its dependency list is used)
    ["org.b:b:1", "fallback"],
  ]);
});

test("unresolvable coordinates are reported once with their requester", async () => {
  const source = new InMemoryPackageSource("test", [
    pkg("org.a:a:1", ["org.gone:gone:9", "org.gone:gone:9"]),
  ]);
  const resolution = await resolveTransitive([c("org.a:a:1")], [source]);
  expect(resolution.missing).toEqual([
    { coordinates: c("org.gone:gone:9"), requestedBy: c("org.a:a:1") },
  ]);
});

test("search merges sources and dedupes by group:artifact", async () => {
  const primary = new InMemoryPackageSource("primary", [pkg("org.x:json-lib:2")]);
  const fallback = new InMemoryPackageSource("fallback", [
    pkg("org.x:json-lib:1"),
    pkg("org.y:json-other:1"),
  ]);
  const hits = await searchPackages("json", [primary, fallback]);
  expect(hits.map(coordinatesToString)).toEqual(["org.x:json-lib:2", "org.y:json-other:1"]);
});

test("latestVersion picks the newest from the first source that knows the package", async () => {
  const source = new InMemoryPackageSource("test", [pkg("org.a:a:1"), pkg("org.a:a:2")]);
  expect(await latestVersion("org.a", "a", [source])).toBe("2");
  expect(await latestVersion("org.nope", "a", [source])).toBeUndefined();
});

File: togo/internal/packages/resolver_test.go
package packages

import (
	"errors"
	"reflect"
	"strings"
	"testing"
)

// dep is a test dependency spec with optional scope/optional flags.
type dep struct {
	spec     string
	scope    string
	optional bool
}

func coord(spec string) Coordinates {
	parts := strings.Split(spec, ":")
	return NewCoordinates(parts[0], parts[1], parts[2])
}

// pkg builds package metadata from a "group:artifact:version" spec and deps.
func pkg(spec string, deps ...dep) PackageMetadata {
	declarations := make([]DependencyDeclaration, 0, len(deps))
	for _, d := range deps {
		declarations = append(declarations, DependencyDeclaration{
			Coordinates: coord(d.spec),
			Scope:       MavenScope(d.scope),
			Optional:    d.optional,
		})
	}
	return PackageMetadata{Coordinates: coord(spec), Dependencies: declarations}
}

func names(packages []ResolvedPackage) []string {
	out := make([]string, len(packages))
	for i, p := range packages {
		out[i] = string(p.Coordinates.String())
	}
	return out
}

// Port of src/packages/resolver.test.ts.

func TestResolveTransitiveBreadthFirst(t *testing.T) {
	source := NewInMemoryPackageSource("test", []PackageMetadata{
		pkg("org.a:a:1", dep{spec: "org.b:b:1"}),
		pkg("org.b:b:1", dep{spec: "org.c:c:1"}),
		pkg("org.c:c:1"),
	})
	res, err := ResolveTransitive([]Coordinates{coord("org.a:a:1")}, []PackageSource{source}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if got := names(res.Packages); !reflect.DeepEqual(got, []string{"org.a:a:1", "org.b:b:1", "org.c:c:1"}) {
		t.Errorf("packages = %v", got)
	}
	depths := []int{res.Packages[0].Depth, res.Packages[1].Depth, res.Packages[2].Depth}
	if !reflect.DeepEqual(depths, []int{0, 1, 2}) {
		t.Errorf("depths = %v", depths)
	}
	if res.Packages[1].RequestedBy != coord("org.a:a:1") {
		t.Errorf("requestedBy = %v", res.Packages[1].RequestedBy)
	}
	if len(res.Conflicts) != 0 || len(res.Missing) != 0 {
		t.Errorf("expected no conflicts/missing, got %v / %v", res.Conflicts, res.Missing)
	}
}

func TestResolveOnResolveOrder(t *testing.T) {
	source := NewInMemoryPackageSource("test", []PackageMetadata{
		pkg("org.a:a:1", dep{spec: "org.b:b:1"}, dep{spec: "org.c:c:1"}),
		pkg("org.b:b:1", dep{spec: "org.c:c:1"}), // c reached twice, resolved once
		pkg("org.c:c:1"),
		pkg("org.x:x:1"),
	})
	var seen []string
	res, err := ResolveTransitive(
		[]Coordinates{coord("org.a:a:1"), coord("org.missing:m:1")},
		[]PackageSource{source},
		func(c Coordinates) { seen = append(seen, string(c.String())) },
	)
	if err != nil {
		t.Fatal(err)
	}
	want := []string{"org.a:a:1", "org.missing:m:1", "org.b:b:1", "org.c:c:1"}
	if !reflect.DeepEqual(seen, want) {
		t.Errorf("seen = %v, want %v", seen, want)
	}
	if len(seen) != len(res.Packages)+len(res.Missing) {
		t.Errorf("seen %d != resolved %d + missing %d", len(seen), len(res.Packages), len(res.Missing))
	}
}

func TestResolveNearestWins(t *testing.T) {
	source := NewInMemoryPackageSource("test", []PackageMetadata{
		pkg("org.a:a:1", dep{spec: "org.b:b:1"}, dep{spec: "org.c:c:1"}),
		pkg("org.b:b:1"),
		pkg("org.b:b:2"),
		pkg("org.c:c:1", dep{spec: "org.b:b:2"}), // farther than a's direct b:1
	})
	res, err := ResolveTransitive([]Coordinates{coord("org.a:a:1")}, []PackageSource{source}, nil)
	if err != nil {
		t.Fatal(err)
	}
	got := names(res.Packages)
	if !contains(got, "org.b:b:1") || contains(got, "org.b:b:2") {
		t.Errorf("expected b:1 selected, b:2 rejected; got %v", got)
	}
	want := []VersionConflict{{Key: "org.b:b", Selected: "1", Rejected: "2", RejectedBy: coord("org.c:c:1")}}
	if !reflect.DeepEqual(res.Conflicts, want) {
		t.Errorf("conflicts = %v, want %v", res.Conflicts, want)
	}
}

func TestResolveCyclesTerminate(t *testing.T) {
	source := NewInMemoryPackageSource("test", []PackageMetadata{
		pkg("org.a:a:1", dep{spec: "org.b:b:1"}),
		pkg("org.b:b:1", dep{spec: "org.a:a:1"}),
	})
	res, err := ResolveTransitive([]Coordinates{coord("org.a:a:1")}, []PackageSource{source}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(res.Packages) != 2 {
		t.Errorf("expected 2 packages, got %d", len(res.Packages))
	}
}

func TestResolveScopeAndOptionalDoNotPropagate(t *testing.T) {
	source := NewInMemoryPackageSource("test", []PackageMetadata{
		pkg("org.a:a:1",
			dep{spec: "org.t:t:1", scope: "test"},
			dep{spec: "org.o:o:1", optional: true},
			dep{spec: "org.r:r:1", scope: "runtime"},
		),
		pkg("org.r:r:1"),
	})
	res, err := ResolveTransitive([]Coordinates{coord("org.a:a:1")}, []PackageSource{source}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if got := names(res.Packages); !reflect.DeepEqual(got, []string{"org.a:a:1", "org.r:r:1"}) {
		t.Errorf("packages = %v", got)
	}
	if len(res.Missing) != 0 {
		t.Errorf("missing = %v", res.Missing)
	}
}

func TestResolveSourcesInOrder(t *testing.T) {
	primary := NewInMemoryPackageSource("primary", []PackageMetadata{pkg("org.a:a:1", dep{spec: "org.b:b:1"})})
	fallback := NewInMemoryPackageSource("fallback", []PackageMetadata{pkg("org.a:a:1"), pkg("org.b:b:1")})
	res, err := ResolveTransitive([]Coordinates{coord("org.a:a:1")}, []PackageSource{primary, fallback}, nil)
	if err != nil {
		t.Fatal(err)
	}
	got := [][2]string{}
	for _, p := range res.Packages {
		got = append(got, [2]string{string(p.Coordinates.String()), string(p.Source)})
	}
	want := [][2]string{{"org.a:a:1", "primary"}, {"org.b:b:1", "fallback"}}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("packages = %v, want %v", got, want)
	}
}

func TestResolveMissingReportedOnce(t *testing.T) {
	source := NewInMemoryPackageSource("test", []PackageMetadata{
		pkg("org.a:a:1", dep{spec: "org.gone:gone:9"}, dep{spec: "org.gone:gone:9"}),
	})
	res, err := ResolveTransitive([]Coordinates{coord("org.a:a:1")}, []PackageSource{source}, nil)
	if err != nil {
		t.Fatal(err)
	}
	want := []MissingPackage{{Coordinates: coord("org.gone:gone:9"), RequestedBy: coord("org.a:a:1")}}
	if !reflect.DeepEqual(res.Missing, want) {
		t.Errorf("missing = %v, want %v", res.Missing, want)
	}
}

func TestSearchMergesSources(t *testing.T) {
	primary := NewInMemoryPackageSource("primary", []PackageMetadata{pkg("org.x:json-lib:2")})
	fallback := NewInMemoryPackageSource("fallback", []PackageMetadata{
		pkg("org.x:json-lib:1"),
		pkg("org.y:json-other:1"),
	})
	hits, err := SearchPackages("json", []PackageSource{primary, fallback})
	if err != nil {
		t.Fatal(err)
	}
	got := make([]string, len(hits))
	for i, h := range hits {
		got[i] = string(h.String())
	}
	if !reflect.DeepEqual(got, []string{"org.x:json-lib:2", "org.y:json-other:1"}) {
		t.Errorf("hits = %v", got)
	}
}

func TestLatestVersion(t *testing.T) {
	source := NewInMemoryPackageSource("test", []PackageMetadata{pkg("org.a:a:1"), pkg("org.a:a:2")})
	if v, _ := LatestVersion("org.a", "a", []PackageSource{source}); v != "2" {
		t.Errorf("LatestVersion = %q, want 2", v)
	}
	if v, _ := LatestVersion("org.nope", "a", []PackageSource{source}); v != "" {
		t.Errorf("LatestVersion(missing) = %q, want empty", v)
	}
}

// erroringSource fails GetMetadata, standing in for a source that exhausted its
// retries on a transient HTTP failure (429/5xx).
type erroringSource struct{ err error }

func (erroringSource) Name() SourceName                              { return "erroring" }
func (erroringSource) Search(string) ([]SearchHit, error)            { return nil, nil }
func (erroringSource) ListVersions(string, string) ([]string, error) { return nil, nil }
func (erroringSource) GetArtifact(Coordinates) ([]byte, error)       { return nil, nil }
func (s erroringSource) GetMetadata(Coordinates) (*PackageMetadata, error) {
	return nil, s.err
}

// Regression for nikeee/cappu#22: a transient fetch failure must abort the
// resolve with the error, never be recorded as a missing package (which the CLI
// would print as "not found in any package source").
func TestResolveTransitivePropagatesFetchError(t *testing.T) {
	boom := errors.New("repo.example: HTTP 429 after 4 attempts")
	res, err := ResolveTransitive([]Coordinates{coord("org.a:a:1")}, []PackageSource{erroringSource{err: boom}}, nil)
	if err == nil {
		t.Fatal("want the fetch error to propagate, got nil")
	}
	if len(res.Missing) != 0 {
		t.Errorf("Missing = %v, want empty (a transient error is not a miss)", res.Missing)
	}
}

func contains(s []string, v string) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}
